ItemProvider -> ItemConvertible, StringRepresentable -> SnakeCaseIdentifiable

liach opened the pull request:
`ItemConvertible` reason:
1. This `ItemProvider` is implemented by `Item` and `Block` instead of some lambda that actually provides/supplies an item.
![image](https://user-images.githubusercontent.com/7806504/56778827-32181700-678d-11e9-8a17-fca58b2275f7.png)
2. Another name `HasItemForm`, which is very accurate, is also considered, but because of the ideal of interface naming in adjectives, I chose `ItemConvertible`, as [`convert` as a word means "change to a different form"](https://www.dictionary.com/browse/convert).

`EnumPropertyValue` reason:
1. This is only implemented by enum classes that can be used as state property values.
![image](https://user-images.githubusercontent.com/7806504/56778869-58d64d80-678d-11e9-8159-451ef3c80cb0.png)
2. This is really just used to get the enum value names in a lowercase form (or getting the proper string version of the value)
![image](https://user-images.githubusercontent.com/7806504/56778975-c84c3d00-678d-11e9-9734-4769f41ca0c2.png)
hence "get value name".
3. A few other usages are present, but they are only really used to be a shortcut for getting lowercase enum names for translation keys, etc.
![image](https://user-images.githubusercontent.com/7806504/56779017-f3369100-678d-11e9-8ef8-61e991fe0eeb.png)

liach committed:
ItemProvider -> ItemConvertible, StringRepresentable -> EnumPropertyValue

Signed-off-by: liach <liach@users.noreply.github.com>

SquidDev commented:
I'm a little unsure about both of these:

 - IMO, `ItemProvider` is perfectly fine. It's never really used in a situation where you care about the "conversion" operation - you just want to get hold of an `Item` instance. If we were to rename it to anything, it'd just be `ItemSupplier`.

 - `StringRepresentable` is a kinda poor name, but `EnumPropertyValue` feels really ugly. Maybe something like `NamedEnum`. Eh, it doesn't feel much better.

liach commented:
1. this item getting interface is only implemented by block and item classes, which aren't providers or suppliers but just have an item form/state.
2. I've considered `PropertyEnum` as well, which means it's an enum used for property (values). How's that? (Might look like `EnumProperty` too much?)

InsomniaKitten commented:
I feel the String supplying interface should be more generalized. Mojang uses it exclusively for determining the names of property values, so the most generalized form of that would be something like `Named` with `getName` as the member name.

liach commented:
Hmm, is there any other string supplying interfaces in vanilla? This one I see are all implemented by enums and used by enum property to retrieve a sanitized name.

Gegy commented on mappings/net/minecraft/item/ItemConvertible.mapping:
```diff
@@ -0,0 +1,2 @@
+CLASS bhg net/minecraft/item/ItemConvertible
```
Like the method, the interface could potentially be named `AsItem` too

liach replied:
hmm, does this match the naming convention? Any other class named similarly in yarn already?

ChloeDawn replied:
I don't think `AsItem` is a very descriptive name. I think `ItemConvertible` is fine.

liach commented:
hmm, does this match the naming convention

ChloeDawn commented on mappings/net/minecraft/item/ItemConvertible.mapping:
```diff
@@ -0,0 +1,2 @@
+CLASS bhg net/minecraft/item/ItemConvertible
+	METHOD g asItem ()Lbbv;
```
`getAsItem` - Following naming convention of Java interfaces that supply an explicitly typed value, such as `BooleanSupplier#getAsBoolean`, and a `get` implies side effects (for example `Block` on the first call looks up the `Item` in the registry and caches it), it is not inherently representing a constant, more a lazy object.

liach replied:
Or `toItem`?

Runemoro commented:
`ItemConvertible` and `StringConvertible` would be better.

InsomniaKitten commented:
`SerializableEnum` might work for the String interface. It is explicitly used for determining a unique name for an enum constant during (de)serialization of state properties.

Runemoro commented:
> `SerializableEnum`

Every enum is serializable (`name` and `byName`).

liach committed:
Name it snake case identifiable instead

Signed-off-by: liach <liach@users.noreply.github.com>

liach commented:
Updated. Renamed `StringRepresentable` to `SnakeCaseIdentifiable` instead, which is better than `StringIdentifiable` (the only method in the interface only return snake case strings)

modmuss50 commented:
Updating, any further discussion needs to go in a new PR or issue.
